crates/viewer/re_time_panel/src/data_density_graph.rs

```diff
@@ -442,6 +428,124 @@ pub fn data_density_graph_ui2(
     }
 }
 
+pub fn build_density_graph<'a>(
+    ui: &'a egui::Ui,
+    time_ranges_ui: &'a TimeRangesUi,
+    row_rect: Rect,
+    db: &re_entity_db::EntityDb,
+    item: &TimePanelItem,
+    timeline: Timeline,
+    config: DensityGraphBuilderConfig,
+) -> DensityGraphBuilder<'a> {
+    re_tracing::profile_function!();
+
+    let mut data = DensityGraphBuilder::new(ui, time_ranges_ui, row_rect);
+
+    // Collect all relevant chunks in the visible time range.
+    // We do this as a separate step so that we can also deduplicate chunks.
+    let visible_time_range = time_ranges_ui
+        .time_range_from_x_range((row_rect.left() - MARGIN_X)..=(row_rect.right() + MARGIN_X));
+
+    // NOTE: These chunks are guaranteed to have data on the current timeline
+    let mut chunk_ranges: Vec<(Arc<Chunk>, ResolvedTimeRange, usize)> = vec![];
+    let mut total_events = 0;
+
+    {
+        visit_relevant_chunks(
+            db,
+            &item.entity_path,
+            item.component_name,
+            timeline,
+            visible_time_range,
+            |chunk, time_range, num_events| {
+                chunk_ranges.push((chunk, time_range, num_events));
+                total_events += num_events;
+            },
+        );
+    }
+
+    // Small chunk heuristics:
+    // We want to render chunks as individual events, but it may be prohibitively expensive
+    // for larger chunks, or if the visible time range contains many chunks.
+    //
+    // We split a large chunk if:
+    // 1. The total number of events is less than some threshold
+    // 2. The number of events in the chunks is less than N, where:
+    //    N is relatively large for sorted chunks
+    //    N is much smaller for unsorted chunks
+
+    {
+        re_tracing::profile_scope!("add_data");
+
+        let can_render_individual_events =
+            config.max_total_chunk_events != 0 && total_events < config.max_total_chunk_events;
```

Comment:
how can an unsigned integer be less than zero?

```suggestion
        let can_render_individual_events = total_events < config.max_total_chunk_events;
```